src/components/RenderCanvas.vue

```diff
@@ -873,22 +935,37 @@ function onRun(runCompiledCode: CompiledPlayground) {
             if (compiler == null) {
                 throw new Error("Could not get compiler");
             }
-            if (compiler.shaderType !== null && !abortRender) {
-                const keepRendering = await execFrame(timeMS, compiler?.shaderType || null, compiledCode, firstFrame);
-                firstFrame = false;
-                return keepRendering;
+            if (abortRender) return false;
+            if (pauseRender.value) return true;
+
+            if (compiler.shaderType === null) {
+                // handle this case
             }
-            return false;
+            const keepRendering = await execFrame(timeMS, compiler?.shaderType || null, compiledCode, firstFrame);
+            firstFrame = false;
+            return keepRendering;
         });
 }
 </script>
 
 <template>
-    <div class="renderOverlay">
-        <div id="performanceInfo">{{ `${frameTime.toFixed(1)} ms` }}</div>
-    </div>
     <canvas v-bind="$attrs" class="renderCanvas" @mousedown="mousedown" @mousemove="mousemove" @mouseup="mouseup"
         ref="canvas"></canvas>
+    <div class="control-bar">
+        <div class="controls-left">
+            <button @click="setFrame(0)" title="Reset frame to 0">&#x23EE;</button>
+            <button @click="setFrame(frameID - 1)" title="Step backward">&#x23F4;</button>
+            <button @click="setFrame(frameID + 1)" title="Step forward">&#x23F5;</button>
+            <button @click="pauseRender = !pauseRender" :title="pauseRender ? 'Resume' : 'Pause'">⏯</button>
+            <span class="frame-counter">Frame: {{ frameID }}</span>
+            <span class="perf-info">{{ frameTime.toFixed(1) }} ms</span>
+            <span class="fps-counter">FPS: {{ fps }}</span>
+            <span class="resolution">Resolution: {{ canvasWidth }}x{{ canvasHeight }}</span>
```

Comment:
```suggestion
            <span class="fps-counter">FPS: {{ Math.round(1000 / frameTime) }}</span>
```

If you calculate this here, it will remove the need to manage it seperately. FPS is a little missleading here though since the true fps is capped at 60.